### PR TITLE
Make source links look cleaner

### PR DESCRIPTION
**Pull request by jsha**
Change from syntaxy-looking [src] to the plain word "source".

Part of #59851

r? @GuillaumeGomez

Demo: https://rustdoc.crud.net/jsha/source-link-2/std/string/struct.String.html

[Discussed on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/display.20of.20source.20link).

### CODE REVIEW

**Comment by camelid** [comment]
Looks good! I don't see the "stable since" change though? Also, the color of "source" looks a little too pale in the light theme.

**Comment by jsha** [comment]
> I don't see the "stable since" change though?

My original version added the "stable since" language but on reviewing the Zulip thread I saw there wasn't really consensus on that, and it's not the main point of the PR, so I reverted it (but forgot to update the PR description). I've now updated the PR description.

> Also, the color of "source" looks a little too pale in the light theme.

I'll bump it up a little.

**Comment by camelid** [comment]
That looks better, although I think the links should be still deeper in color. The current color in some ways feels more distracting than if it were darker because it's on the edge of being less visible vs more visible. It just doesn't look right to me for whatever reason.

**Comment by jsha** [comment]
Just pushed a version where they get normal opacity like any other link. Lemme know what you think.

**Comment by camelid** [comment]
That looks better, thanks.

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -312,15 +312,15 @@ impl AllTypes {
...
-                             [<span class=\"inner\">&#x2212;</span>]\
-                         </a>\
-                     </span>
-                 </span>
-             </h1>",
+             </h1>\
+             <span class=\"out-of-band\">\
+                <span id=\"render-detail\">\
+                    <a id=\"toggle-all-docs\" href=\"javascript:void(0)\" \
+                    title=\"collapse all docs\">\
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:319]: nit: could could indent the `title` attribute so my brain doesn't interpret it as a tag please?

**Comment by GuillaumeGomez** [comment]
Apart from my comment, looks good to me.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
So looks good to me. I see some :-1: signs on the first comment.

@euclio @falaosu: What don't you like with this approach?

**Comment by euclio** [comment]
My original objection was that the new text isn't visually distinct enough, though after reading the linked issue I see that there's a justification for this.

That said, I don't think that this does a good enough job of communicating that these are clickable actions. The brackets in the current style do a good job of making it clear that these are buttons, and "-" communicates the same amount of information as "collapse" in fewer characters. For such commonly-used actions that are repeated across the page, we can use symbols to communicate these actions effectively, and that reduces overall clutter.

"source" is maybe fine as a link visually, since it actually takes you to a separate page (though there's no underline on hover). I don't think collapse should just be a link, though, since it only performs an action on the current page. Assuming that we want to change the other "[-]"s on the page for consistency, I would expect a symbol like "▶︎" for each individual section, and "▼ collapse all" at the top.

**Comment by jsha** [comment]
> "source" is maybe fine as a link visually, since it actually takes you to a separate page (though there's no underline on hover).

I'll add underline on hover.

> I don't think collapse should just be a link, though, since it only performs an action on the current page. Assuming that we want to change the other "[-]"s on the page for consistency, I would expect a symbol like "▶︎" for each individual section, and "▼ collapse all" at the top.

I changed `[-]` to `collapse` because `1.0.0 · [-] · source` looked wrong. But I could make it `1.0.0 · source · [-]`, in the spirit of "change one thing at a time."

**Comment by GuillaumeGomez** [comment]
> "source" is maybe fine as a link visually, since it actually takes you to a separate page (though there's no underline on hover). I don't think collapse should just be a link, though, since it only performs an action on the current page. Assuming that we want to change the other "[-]"s on the page for consistency, I would expect a symbol like "▶︎" for each individual section, and "▼ collapse all" at the top.

To be fair, this button at the top of the page is a bit on its own since its behaviour is different than the other collapse toggles. But at least I see your point.

@jsha: Just realized that you didn't add a test for the text of the top collapse button (to check it's "expand"/"collapse" when it should). Could you add one as well please? And since you said you would add underline on hover, can you add a test for this as well please?

**Comment by jsha** [comment]
Alright, I've pushed new code reverting the "collapse" text back to `[-]`. Updated the demo.

> Just realized that you didn't add a test for the text of the top collapse button (to check it's "expand"/"collapse" when it should).

Since this is reverted, no behavior here has changed. I'll look at adding a test for the `+` / `-` change though.

> And since you said you would add underline on hover, can you add a test for this as well please?

Will do.

**Comment by camelid** [comment]
FWIW, I liked the `[-]` -> `collapse` change. But, in the interest of incremental progress, I think it's fine to discuss that later.

**Comment by GuillaumeGomez** [comment]
I liked it as well, maybe for another time...

**Comment by jsha** [comment]
@GuillaumeGomez I've added the tests.

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -312,15 +312,15 @@ impl AllTypes {
...
-                 <span class=\"out-of-band\">\
-                     <span id=\"render-detail\">\
-                         <a id=\"toggle-all-docs\" href=\"javascript:void(0)\" \
-                            title=\"collapse all docs\">\
-                             [<span class=\"inner\">&#x2212;</span>]\
-                         </a>\
-                     </span>
-                 </span>
-             </h1>",
+             </h1>\
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:315]: Why did you change this? With this change it looks like this:

Screenshot from 2022-01-07 21-49-05

**Comment by GuillaumeGomez** [comment]
Thanks for the tests! Just one last question and it's good for me. :)

File: src/librustdoc/html/render/mod.rs (hunk 2 above)
Comment (jsha) [reply src/librustdoc/html/render/mod.rs]: Good catch! This was a mistake from an earlier iteration. Reverted.

File: src/librustdoc/html/render/mod.rs [hunk 3]
```diff
@@ -315,8 +315,8 @@ impl AllTypes {
                  <span class=\"out-of-band\">\
                      <span id=\"render-detail\">\
                          <a id=\"toggle-all-docs\" href=\"javascript:void(0)\" \
-                            title=\"collapse all docs\">\
-                             [<span class=\"inner\">&#x2212;</span>]\
+                             title=\"collapse all docs\">\
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:318]: Why these changes?

**Comment by GuillaumeGomez** [comment]
When I hover the source links, they don't get underlined. Could fix it and add a test for it too please?

**Comment by jsha** [comment]
> When I hover the source links, they don't get underlined. Could fix it and add a test for it too please?

The source link at the top of the page is part of the main-heading, where links get underline on hover.

The source link to the right of the code headings is part of those headings, which don't get underline on hover.

I think this PR is consistent with how we currently do things. If we want to add underline-on-hover everywhere, I'm in favor, but we shouldn't do it piecemeal.

I pushed to remove the spurious diffs. Thanks for catching!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
It's fine as is then. I think it's worth discussing for a follow-up which links should be underlined or not but it's for later. r=me once CI pass and thanks!
